sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1134,28 +1119,44 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
   CopyCommandBatch.QueueBatchSize = ZeCommandListBatchCopyConfig.startSize();
 }
 
+static pi_result cleanup(pi_event);
+
 // Reset signalled command lists in the queue and put them to the cache of
 // command lists. A caller must not lock the queue mutex.
-pi_result _pi_queue::resetCommandLists() {
-  // We check for command lists that have been already signalled, but have not
-  // been added to the available list yet. Each command list has a fence
-  // associated which tracks if a command list has completed dispatch of its
-  // commands and is ready for reuse. If a command list is found to have been
-  // signalled, then the command list & fence are reset and command list is
-  // returned to the command list cache. All events associated with command list
-  // are cleaned up if command list was reset.
-  std::scoped_lock Lock(this->Mutex);
-  for (auto &&it = CommandListMap.begin(); it != CommandListMap.end(); ++it) {
-    // It is possible that the fence was already noted as signalled and
-    // reset. In that case the InUse flag will be false.
-    if (it->second.InUse) {
-      ze_result_t ZeResult =
-          ZE_CALL_NOCHECK(zeFenceQueryStatus, (it->second.ZeFence));
-      if (ZeResult == ZE_RESULT_SUCCESS) {
-        PI_CALL(resetCommandList(it, true));
+pi_result resetCommandLists(pi_queue Queue) {
+  std::vector<pi_event> EventListToCleanup;
+  {
+    // We check for command lists that have been already signalled, but have not
+    // been added to the available list yet. Each command list has a fence
+    // associated which tracks if a command list has completed dispatch of its
+    // commands and is ready for reuse. If a command list is found to have been
+    // signalled, then the command list & fence are reset and command list is
+    // returned to the command list cache. All events associated with command
+    // list are cleaned up if command list was reset.
+    std::scoped_lock Lock(Queue->Mutex);
+    for (auto &&it = Queue->CommandListMap.begin();
+         it != Queue->CommandListMap.end(); ++it) {
+      // It is possible that the fence was already noted as signalled and
+      // reset. In that case the InUse flag will be false.
+      if (it->second.InUse) {
+        ze_result_t ZeResult =
+            ZE_CALL_NOCHECK(zeFenceQueryStatus, (it->second.ZeFence));
+        if (ZeResult == ZE_RESULT_SUCCESS) {
+          PI_CALL(Queue->resetCommandList(it, true, EventListToCleanup));
+        }
       }
     }
   }
+  for (auto Event : EventListToCleanup) {
+    // We don't need to synchronize the events since the fence
+    // synchronized above already does that.
+    Event->Completed = true;
+
+    PI_CALL(cleanup(Event));
+    // These events were just removed from command list, so decrement ref count
+    // (it was incremented when they were added to the command list).
+    PI_CALL(piEventRelease(Event));
```

Comment:
Why did you move this out of cleanup()?

Reply:
This piEventRelease is moved here from resetCommandList, not from cleanup. We reset command list above  and removed these events from its EventList container. So we can decrement refcount.
The idea was to move cleanup and piEventRelease functions out of queue lock because they can possibly call piQueueRelease which locks the queue. 
I can move piEventRelease back to the resetCommandList if https://github.com/intel/llvm/pull/6192 gets merged.
So it is probably better to merge https://github.com/intel/llvm/pull/6192  first, then I can update this patch.

Reply:
I see, thanks. I think it is clearer if we release goes in the same place where the reference is removed.